pilot/pkg/proxy/envoy/v2/debug.go

```diff
@@ -244,3 +267,138 @@ func (s *DiscoveryServer) registryz(w http.ResponseWriter, req *http.Request) {
 		_, _ = w.Write(b)
 	}
 }
+
+// Endpoint debugging
+func (s *DiscoveryServer) endpointz(w http.ResponseWriter, req *http.Request) {
+	_ = req.ParseForm()
+	svcName := req.Form.Get("svc")
+	if svcName != "" {
+		data, err := ioutil.ReadAll(req.Body)
+		if err != nil {
+			return
+		}
+		svc := &model.ServiceInstance{}
+		err = json.Unmarshal(data, svc)
+		if err != nil {
+			return
+		}
+		s.MemRegistry.AddInstance(svcName, svc)
+	}
+	brief := req.Form.Get("brief")
+	if brief != "" {
+		svc, _ := s.env.ServiceDiscovery.Services()
+		for _, ss := range svc {
+			for _, p := range ss.Ports {
+				all, err := s.env.ServiceDiscovery.Instances(ss.Hostname, []string{p.Name}, nil)
+				if err != nil {
+					return
+				}
+				for _, svc := range all {
+					fmt.Fprintf(w, "%s:%s %s:%d %v %v %s\n", ss.Hostname,
+						p.Name, svc.Endpoint.Address, svc.Endpoint.Port, svc.Labels,
+						svc.AvailabilityZone, svc.ServiceAccount)
+				}
+			}
+		}
+		return
+	}
+
+	svc, _ := s.env.ServiceDiscovery.Services()
+	fmt.Fprint(w, "[\n")
+	for _, ss := range svc {
+		for _, p := range ss.Ports {
+			all, err := s.env.ServiceDiscovery.Instances(ss.Hostname, []string{p.Name}, nil)
+			if err != nil {
+				return
+			}
+			fmt.Fprintf(w, "\n{\"svc\": \"%s:%s\", \"ep\": [\n", ss.Hostname, p.Name)
+			for _, svc := range all {
+				b, err := json.MarshalIndent(svc, "  ", "  ")
+				if err != nil {
+					return
+				}
+				_, _ = w.Write(b)
+				fmt.Fprint(w, ",\n")
+			}
+			fmt.Fprint(w, "\n{}],")
+		}
+	}
+	fmt.Fprint(w, "\n{}]\n")
+}
+
+// Config debugging.
+func (s *DiscoveryServer) configz(w http.ResponseWriter, req *http.Request) {
+	fmt.Fprintf(w, "\n[\n")
+	for _, typ := range s.env.IstioConfigStore.ConfigDescriptor() {
+		cfg, _ := s.env.IstioConfigStore.List(typ.Type, "")
+		for _, c := range cfg {
+			b, err := json.MarshalIndent(c, "  ", "  ")
+			if err != nil {
+				return
+			}
+			_, _ = w.Write(b)
+			fmt.Fprint(w, ",\n")
+		}
+	}
+	fmt.Fprint(w, "\n{}]")
+}
+
+// adsz implements a status and debug interface for ADS.
+// It is mapped to /debug/adsz
+func adsz(w http.ResponseWriter, req *http.Request) {
+	_ = req.ParseForm()
+	if req.Form.Get("debug") != "" {
+		adsDebug = req.Form.Get("debug") == "1"
+		return
+	}
+	if req.Form.Get("push") != "" {
+		adsPushAll()
+		fmt.Fprintf(w, "Pushed to %d servers", len(adsClients))
+		return
+	}
+	adsClientsMutex.RLock()
+
+	//data, err := json.Marshal(ldsClients)
+
+	// Dirty json generation - because standard json is dirty (struct madness)
+	// Unfortunately we must use the jsonbp to encode part of the json - I'm sure there are
+	// better ways, but this is mainly for debugging.
+	fmt.Fprint(w, "[\n")
+	comma2 := false
+	for _, c := range adsClients {
```

Comment:
Golang has built-in struct-to-json marshaling, which does it correctly. Just add json field names to the XdsConnection and call `json.Marshall`. You can even marshal the entire map in one go if both the keys and values can be marshaled: `json.Marshal(adsClients)`. https://gobyexample.com/json

Reply:
And further, Proto-generated structs also include json field name tags, so they can be marshalled into (non-Proto) JSON by calling `json.Marshal` too.

Reply:
I'm well aware of this - but it doesn't work well with 'structs' - the generated output is a mess. 
I wouldn't have wasted the time doing this if the simpler solution worked...

Reply:
Ug, that's frustrating (given that `struct` is supposed to be JSON directly).